### ISSUE TITLE
Redux forgets facts involving nonlinear multiplication

### ISSUE THREAD

**Issue by necto** [verifast/verifast#17]
I am in the process of proving a simple arithmetic property. But after I apply a lemma, VeriFast is unable to prove the assertion it verified one line ago. Here `a == ((a/b)*b) + (a%b)` is provided by the `div_rem` lemma, and it verifies in the first assert, but fails to verify in the second one - after I add a `0 == (a/b)` fact to the assumptions stack.

```
/*@
lemma void division_rounds_to_zero(int a, int b);
requires 0 <= a &*& a < b;
ensures 0 == (a/b);

lemma void to_prove(int a, int b)
requires 0 <= a &*& a < b;
ensures a == (a%b);
{
   div_rem(a,b);
   assert(a == ((a/b)*b) + (a%b)); //<-- This is OK.
   division_rounds_to_zero(a,b);
   assert(a == ((a/b)*b) + (a%b)); // <-- This one fails.
}
@*/
```

:arrow_right:  _Assertion might not hold: a = ((/(a, b) \* b) + %(a, b))_

**Comment by btj** [comment]
This is a Redux problem; the example works fine with `-prover z3v4.5`.

A simpler case is the following:
```c
/*@

lemma void redux_mul_incompleteness(real a, real b, real c, real d)
    requires a == b * c &*& b == 0;
    ensures a == b * c; // fails
{
}

@*/
```
The problem is that Redux treats the first occurrence of `b * c` as an application of an uninterpreted function symbol, but it simplifies the second one to just 0 because at that point it knows that `b = 0`.

The way to fix this is probably to check, whenever a value node is merged with a literal, if it has any `mul` parents and to merge these parents with the resolved polynomial. That is, when Redux learns that `b = 0`, it should merge term `b * c` with 0.

**Comment by btj** [comment]
Fixed by f90964f51ffb33bdf3d1ccd9c0d473fe32e2955a .
